Angus/1259 multi panel

veggiesaurus opened the pull request:
Closes #1259. Some bugs will be solved in separate issues. The main aim of this PR is to add an experimental multi-panel view without breaking the existing single-panel view.

- [x] Handle image export in multi-panel view.
- [x] Required tiles are sometimes not updated when zooming in, resulting in pixelated images.
- [x] Beam profile overlays not implemented correctly yet
- [x] The image toolbar should only show up on one panel at a time, and be aligned correctly
- [x] Catalogs should be rendered correctly
- [x] Colorbar's colormap and values should be determined per-image
- [x] Active panel highlight needs to be adjusted to be more obvious
- [x] Disable columns/rows spinner when multi-panel is disabled. Adjust text when in dynamic grid mode
- [x] Align non-ideal state 
- [x] Images sometimes don't show up if you load them simultaneously.

veggiesaurus committed:
introduce image panel component

veggiesaurus committed:
moved resize detector to image view component

veggiesaurus committed:
initial support for panels (raster and contour broken)

veggiesaurus committed:
dynamic grid sizes

veggiesaurus committed:
dynamic grid sizes

veggiesaurus committed:
cursor info shown based on preferences

veggiesaurus committed:
fixed issue with loading files from URL

veggiesaurus committed:
formatting pass

veggiesaurus committed:
fixed issue with initial zoom levels

veggiesaurus committed:
minor formatting change

veggiesaurus committed:
determine current page by active image index

veggiesaurus committed:
Merge branch 'dev' into angus/1259_multi-panel

veggiesaurus committed:
Merge branch 'dev' into angus/1259_multi-panel

veggiesaurus committed:
basic grid rendering of raster components

veggiesaurus committed:
grid rendering of contours

veggiesaurus committed:
minor cleanup

veggiesaurus committed:
switch active image on interaction (pan/zoom/click)

veggiesaurus committed:
highlight active image (by doubling AST border) when in tiled view

veggiesaurus committed:
ensure pixel coordinates are plotted when frame has no WCS information

veggiesaurus committed:
moved image ratio popup to image view component

veggiesaurus committed:
added tiling to image view config as silent prefs

veggiesaurus committed:
adjusted preference names

veggiesaurus committed:
formatting pass

kswang1029 commented:
@veggiesaurus I have done some tests. The following issues/comments are what I found with the current implementation state. **_Not all need to be addressed for v3b1._** 

1. some (or all?) image view settings (eg title, colorbar's colormap/value) need to be generalized to support multiple panels. That is, each image would have its own settings instead of having a global one applied to all images. The UI needs to be revised carefully so that users know which active panel will be configured. KSW's homework...
Title and colorbar are more important for now as they have usability impact. 

2. Depending on the grid configuration, the toolbar at the bottom-right corner of each panel sometimes does not have sufficient space to display in the horizontal direction. KSW's homework on the solution...

3. The active panel highlight is less noticeable. Can we highlight the border of "image-panel-div" instead?

4. If multi-panel mode = "dynamic grid size" and we load two images in one shot via the file browser, the left panel will be empty (raster image not displayed). If we load the two images one by one, then it works fine. Cache issue?

5. In the image view settings, shall we show panel columns/rows spinbox only when multi-panel mode is "fixed grid size"?

6. Depending on which active image it is, the catalog overlays on different images are not rendered correctly (they are scaled). In the single panel mode, it works fine. 

https://user-images.githubusercontent.com/20819712/128120256-f01d500f-df38-4dea-83c3-9b1a41a8749a.mov


7. Tested 4x4 as performance test. We can "feel" delay in both pan and zoom actions but I think this is okay for now.

8. When dynamic grid is default and we load multiple images in one shot, sometimes we see tiles are misplaced in a panel. This displayed content changes with different zoom levels. This likely happens when the loaded images have different sizes in pixels. 
![Screen Shot 2021-08-04 at 11 57 52 AM](https://user-images.githubusercontent.com/20819712/128120037-2a8d3aff-f8eb-4637-bf33-7fb0523dbd3d.png)

9. The non-ideal state "No file loaded" of the image viewer is shifted upward. 

![Screen Shot 2021-08-04 at 12 02 10 PM](https://user-images.githubusercontent.com/20819712/128120352-8ef42259-3c55-4965-bac1-bead3c4300b0.png)

10. The missing tile problem could occur when loading multiple images in one shot with the dynamic grid mode. Example in the screenshot of 8)

11.  Ideally, when we show two matched cubes side by side, change the channel in one should also trigger image update of the other. As for animation, if performance allows and there is no other technical limitation, the behavior should be the same.

12. [Comment] The combination of multi-panel view and the multi-profile view in the spectral profile widget provides very nice UX!
![Screen Shot 2021-08-04 at 12 19 50 PM](https://user-images.githubusercontent.com/20819712/128121595-2654c344-4010-46c4-9ae8-24f5af81661a.png)

veggiesaurus committed:
Merge branch 'dev' into angus/1259_multi-panel

veggiesaurus commented:
@kswang1029 here are my responses and suggestions in terms of what gets handled in beta-1:

> 1. some (or all?) image view settings (eg title, colorbar's colormap/value) need to be generalized to support multiple panels. That is, each image would have its own settings instead of having a global one applied to all images. The UI needs to be revised carefully so that users know which active panel will be configured. KSW's homework...

I think the colorbar's colormap/value should be an easy fix if I refactor things a bit to make it associated with the _frame_ rather than the overlay store. I'll try do that for beta-1. The title stuff is a little more complicated, because we currently have a single active WCS system, so all images that have WCS are forced to use the same system as the active image. The only way to prevent this from happening is actually to make the active image one without WCS, and each image reverts to its native system:
![image](https://user-images.githubusercontent.com/592504/128166272-10588af5-fe8e-45c4-94c9-c304a7b675c5.png)

Custom titles per image will be more difficult and will need a bit of a rethink in terms of UI, so we should add those to beta-2.

> 2. Depending on the grid configuration, the toolbar at the bottom-right corner of each panel sometimes does not have sufficient space to display in the horizontal direction. KSW's homework on the solution...

I'd like the toolbar to only show in a single image at a time. If the image is in the left-most column, then we can align it left, rather than right. e.g:
![image](https://user-images.githubusercontent.com/592504/128167069-b58f4b27-4c69-4371-9b64-55c606154132.png)
I can add this feature to beta-1. 

> 3. The active panel highlight is less noticeable. Can we highlight the border of "image-panel-div" instead?

Good idea. I'll do this in beta-1.

> 4. If multi-panel mode = "dynamic grid size" and we load two images in one shot via the file browser, the left panel will be 
empty (raster image not displayed). If we load the two images one by one, then it works fine. Cache issue?

Will try to fix this is beta-1. It's got to do with the initial view size, I think. As soon as you move the image it seems to show up again.

> 5. In the image view settings, shall we show panel columns/rows spinbox only when multi-panel mode is "fixed grid size"?

We need to show this when it's in dynamic grid size too, as  this defines the maximum columns. Perhaps I can change the text to be "Max panel columns" when you switch to dynamic grid, and disable when in single-panel. Can fix for beta-1

> 6. Depending on which active image it is, the catalog overlays on different images are not rendered correctly (they are scaled). In the single panel mode, it works fine.

I will try fix this for beta-1. I haven't currently attempted to get catalogs to work in multipanel mode.

> 1. Tested 4x4 as performance test. We can "feel" delay in both pan and zoom actions but I think this is okay for now.

Not sure how to get round this other than severely throttling render redraw when images are not active and the user is panning/zooming.

> 2. When dynamic grid is default and we load multiple images in one shot, sometimes we see tiles are misplaced in a panel. This displayed content changes with different zoom levels. This likely happens when the loaded images have different sizes in pixels.

I noticed that too. Not entirely sure what's going on. I'll try solve this. Does it ever occur if you load images one-by-one?

> 3. The non-ideal state "No file loaded" of the image viewer is shifted upward.

Will fix in beta-1

> 1. The missing tile problem could occur when loading multiple images in one shot with the dynamic grid mode. Example in the screenshot of 8)
> 2. Ideally, when we show two matched cubes side by side, change the channel in one should also trigger image update of the other. As for animation, if performance allows and there is no other technical limitation, the behavior should be the same.

I wonder if this is related to #1571. Do contours update correctly?

veggiesaurus committed:
realign NonIdealState div

veggiesaurus committed:
nicer handling of panel highlighting when image is active

veggiesaurus committed:
disable row and column spinners in single-panel mode, changed labels

veggiesaurus committed:
removed raster render type

veggiesaurus committed:
only show toolbar on one image at a time

veggiesaurus committed:
auto-fit when image view is resized from initial 1x1 placeholder size

veggiesaurus committed:
render correct catlogs for each visible image

veggiesaurus committed:
use current frame instead of active frame when determining catalog control maps

veggiesaurus committed:
display correct beams for each visible image

veggiesaurus committed:
separate ticks numbers and positions for each colorbar

veggiesaurus committed:
combine all images when exporting

veggiesaurus commented:
Ready for review. Will solve the outstanding issue above ("Required tiles are sometimes not updated when zooming in, resulting in pixelated images.") in a separate issue.

veggiesaurus committed:
offset image by 1px to fix compositing bug

veggiesaurus committed:
round window height down to prevent offset bugs

veggiesaurus committed:
fix issue with panning and zooming tiled images

veggiesaurus committed:
fix issue with panning and zooming tiled images

veggiesaurus committed:
fixed typo leading to aggressive clearing of pending request map

veggiesaurus committed:
fixed mobx warning on mouse hover

veggiesaurus committed:
Merge branch 'dev' into angus/1259_multi-panel

kswang1029 commented:
- [ ] the pixel highlight (mouse on colorbar) feature does not work properly on the non-active panel. The threshold value is adopted from the active panel but then applied to the non-active panel.

https://user-images.githubusercontent.com/20819712/128654330-7b33ca5b-a5ae-4604-996a-6b5cab3475ab.mov

- [ ]  when loading multiple images in one shot, some images do not show full FOV in the panels (all except the bottom-right panel). The zoom levels seem to change panel by panel 🤔. I guess this is because of the dynamic grid generation? The final 2x2 grid is actually generated from 1x1 to 1x2, then 2x2. It would be good if this can be fixed but it is probably okay to keep this "behavior". (PS. missing tiles could happen too 🙈)

https://user-images.githubusercontent.com/20819712/128654595-15ee31ea-dfb1-4025-bb2e-d13a410b6ff6.mov

 - [ ] with a Retina display, the exported png seems to have the deviceRatio issue. What we see on screen is
 
<img width="1071" alt="Screen Shot 2021-08-09 at 10 35 01 AM" src="https://user-images.githubusercontent.com/20819712/128655160-f5b152dd-2cd7-4ec5-875d-615cd61dda43.png">

but the exported png is
![dice_one fits-dice_two fits-dice_three fits-dice_four fits-image-2021-08-09-10-34-34](https://user-images.githubusercontent.com/20819712/128655188-785fdc50-c2f6-46b2-9347-a255d683bce4.png)

kswang1029 commented:
As a note here: one time I see a panel contains disordered tiles at different zoom levels. It cannot be reproduced so far. I will keep an eye on this.

YuHsuan-Hwang commented on src/stores/OverlayStore.ts:
```diff
@@ -973,7 +910,11 @@ export class OverlayColorbarSettings {
+        if (!activeFrame) {
+            return 0;
+        }
+        const textWidth = Math.max(...activeFrame.colorbarStore.texts.map(x => x.length)) * this.textRatio[clamp(Math.floor(this.numberFont / 4), 0, this.textRatio.length)];
```
The changes about colorbar looks good to me 👍 
Just found one possible enhancement: we can use the max of `Math.max(...activeFrame.colorbarStore.texts.map(x => x.length))` of all 1~4 frames, instead of just checking the active frame. The `numberWidth` will not be changing when we click different panels in this way

YuHsuan-Hwang commented:
Minor issues of the toolbar:
The toolbar is placed under the colorbar. may be a z-index issue
We also can not click the buttons if it overlaps with another toolbar
![image](https://user-images.githubusercontent.com/43841102/128676851-ce3bf574-e93e-4949-9aac-9115604d0051.png)

veggiesaurus committed:
auto-fit loaded images after all are loaded

veggiesaurus commented:
> * [ ]   when loading multiple images in one shot, some images do not show full FOV in the panels (all except the bottom-right panel). The zoom levels seem to change panel by panel thinking. I guess this is because of the dynamic grid generation? The final 2x2 grid is actually generated from 1x1 to 1x2, then 2x2. It would be good if this can be fixed but it is probably okay to keep this "behavior". (PS. missing tiles could happen too see_no_evil)

Yes, it's because of the dynamic grid size. Fixed in `05a78c7`.

veggiesaurus committed:
fixed image export with retina screens

veggiesaurus committed:
fixed colorbar value on non-active visible frames

veggiesaurus committed:
overlay numberWidth determined from all visible images

veggiesaurus committed:
remove left-alignment of toolbar

veggiesaurus commented:
@kswang1029, I've fixed the issues you raised above. Please re-check. @YuHsuan-Hwang I've implemented the `numberWidth` suggestion you made as well. 

As for the toolbar: it's problematic. I'm not sure how to solve the issue properly. Things behave badly once the toolbar is too long. I've removed the left-alignment for now. I think the only way to really solve this is to have some sort of responsive design that limits the toolbar to a certain width based on the panel size. But I'd rather keep that for another PR. Can we mark it as a know issue for now?

kswang1029 commented:
> @kswang1029, I've fixed the issues you raised above. Please re-check. @YuHsuan-Hwang I've implemented the `numberWidth` suggestion you made as well.
> 
> As for the toolbar: it's problematic. I'm not sure how to solve the issue properly. Things behave badly once the toolbar is too long. I've removed the left-alignment for now. I think the only way to really solve this is to have some sort of responsive design that limits the toolbar to a certain width based on the panel size. But I'd rather keep that for another PR. Can we mark it as a know issue for now?

Sure for another PR. It could be that if the space is not enough, we show a cog or something instead and user can hover over it to expand all the buttons. Surely we need a 2D grid of buttons I guess

veggiesaurus commented:
> It could be that if the space is not enough, we show a cog or something instead and user can hover over it to expand all the buttons.

That's what I was thinking :+1:

kswang1029 commented:
> > It could be that if the space is not enough, we show a cog or something instead and user can hover over it to expand all the buttons.
> 
> That's what I was thinking 👍

Alternatively when the grid mode is on, we reserve a narrow horizontal space at the bottom of the viewer and put all the buttons there. They apply to the active panel only.

veggiesaurus commented:
@kswang1029 @YuHsuan-Hwang what about this approach (using a flex wrap):
![image](https://user-images.githubusercontent.com/592504/128725144-bb55107d-c62a-46c2-936c-28a476d35709.png)

turns into this when resizing:
![image](https://user-images.githubusercontent.com/592504/128725264-d5d11184-a4ad-454d-96ad-4dcb2e002722.png)

or if necessary: 
![image](https://user-images.githubusercontent.com/592504/128725443-8f4ecfd4-4e06-4b27-918f-be568a941d49.png)

A more realistic scenario:
![image](https://user-images.githubusercontent.com/592504/128725817-834c7333-b9d8-442f-9b0f-08029a841a21.png)


It's still entirely usable when it wraps across to two lines. I think this seems like a nice solution. We could also shrink the toolbar a bit:

![image](https://user-images.githubusercontent.com/592504/128727334-fa9c37fe-04b0-4d87-8fde-ae5a12bb4aef.png)

but I think that looks a little clumsy.

veggiesaurus committed:
flex wrap toolbar when panel is too small

kswang1029 commented:
> @kswang1029 @YuHsuan-Hwang what about this approach (using a flex wrap):
> ![image](https://user-images.githubusercontent.com/592504/128725144-bb55107d-c62a-46c2-936c-28a476d35709.png)
> 
> turns into this when resizing:
> ![image](https://user-images.githubusercontent.com/592504/128725264-d5d11184-a4ad-454d-96ad-4dcb2e002722.png)
> 
> or if necessary:
> ![image](https://user-images.githubusercontent.com/592504/128725443-8f4ecfd4-4e06-4b27-918f-be568a941d49.png)
> 
> A more realistic scenario:
> ![image](https://user-images.githubusercontent.com/592504/128725817-834c7333-b9d8-442f-9b0f-08029a841a21.png)
> 
> It's still entirely usable when it wraps across to two lines. I think this seems like a nice solution. We could also shrink the toolbar a bit:
> 
> ![image](https://user-images.githubusercontent.com/592504/128727334-fa9c37fe-04b0-4d87-8fde-ae5a12bb4aef.png)
> 
> but I think that looks a little clumsy.

yes this seems better. However it looks a bit less pleasant due to misalignment. Ok for beta1 for now. We may consider the following approach for beta2. 

https://user-images.githubusercontent.com/20819712/128797306-8ff13feb-9d19-46cf-8bf0-0ca9f414bf5c.mov

The idea is when the space is not enough, we show a button instead and by hovering it all the buttons show up in a popover. This would make the layout cleaner I guess.

kswang1029 reviewed (approved):
Another milestone for the ultimate CARTA 👍
